File: php/array_based_stack.php
<?php
class Stack
{
    private $stack; // Array to store stack elements
    private $size; // Number of elements in the stack

    public function __construct()
    {
        $this->stack = []; // Initialize an empty array
        $this->size = 0; // Stack size is initially 0
    }

    public function push($element)
    {
        $this->stack[] = $element; // Add element to the end of the array
        $this->size++; // Increment the size
        echo "Pushed: $element\n";
    }

    public function pop()
    {
        if ($this->isEmpty())
        {
            echo "Stack is empty. Cannot pop.\n";
            return null;
        }

        $poppedElement = array_pop($this->stack); // Remove the last element
        $this->size--; // Decrement the size
        echo "Popped: $poppedElement\n";
        return $poppedElement;
    }

    public function peek()
    {
        if ($this->isEmpty())
        {
            echo "Stack is empty. Nothing to peek\n";
            return null;
        }

        return $this->stack[$this->size - 1]; // Return the last element
    }

    public function isEmpty()
    {
        return $this->size === 0; // Returns true if size is 0
    }

    public function getSize()
    {
        return $this->size;
    }

    public function display()
    {
        if ($this->isEmpty())
        {
            echo "Stack is empty.\n";
            return;
        }

        echo "Stack elements (bottom to top):\n";
        foreach ($this->stack as $element)
        {
            echo $element . "\n";
        }
    }

    public function clear()
    {
        $this->stack = []; // Reset the stack to an empty array
        $this->size = 0; // Reset the size to 0
        echo "Stack cleared.\n";
    }
}

$stack = new Stack();

$stack->push(10);
$stack->push(20);
$stack->push(30);

$stack->display(); // Output: 10, 20, 30

echo "Top element: " . $stack->peek() . PHP_EOL; // Output: 30

$stack->pop();
$stack->display(); // Output: 10, 20

echo "Stack size: " . $stack->getSize() . PHP_EOL; // Output: 2

$stack->clear();
$stack->display(); // Output: Stack is empty.
?>

File: kotlin/src/ArrayBasedStack.kt
class ArrayStack<T>(private val capacity: Int)
{
    private var stack = arrayOfNulls<Any>(capacity) // Array to store elements
    private var top = -1 // Track the top of the stack (-1 means empty)

    fun push(value: T)
    {
        if (top == capacity - 1)
        {
            println("Stack Overflow! Cannot add $value")
            return
        }
        top++
        stack[top] = value
        println("Pushed $value onto the stack.")
    }

    fun pop(): T?
    {
        if (isEmpty())
        {
            println("Stack Underflow! Stack is empty.")
            return null
        }
        val poppedValue = stack[top]
        stack[top] = null // Remove reference for garbage collection
        top--
        println("Popped $poppedValue from the stack.")
        return poppedValue as T?
    }

    fun peek(): T?
    {
        return if (isEmpty())
        {
            println("Stack is empty.")
            null
        }
        else
        {
            println("Top of the stack is: ${stack[top]}")
            stack[top] as T
        }
    }

    fun isEmpty(): Boolean
    {
        return top == -1
    }

    fun size(): Int
    {
        return top + 1
    }

    fun clear()
    {
        for (i in 0..top)
        {
            stack[i] = null
        }
        top = -1
        println("Stack cleared.")
    }

    fun printStack()
    {
        if (isEmpty())
        {
            println("Stack is empty.")
        }
        else
        {
            println("Stack elements from top to bottom:")
            for (i in top downTo 0)
            {
                println(stack[i])
            }
        }
    }
}

fun main()
{
    val stack = ArrayStack<Int>(5)

    println("\n-- Pushing Elements onto the Stack --")
    stack.push(10)
    stack.push(20)
    stack.push(30)
    stack.push(40)
    stack.push(50)
    stack.push(60) // Stack overflow

    println("\n-- Popping Elements from the Stack --")
    stack.pop() // Should remove 50
    stack.pop() // Should remove 40

    println("\n-- Current Stack Size --")
    println("Stack size: ${stack.size()}") // Should print 3

    println("\n-- Printing Stack Elements --")
    stack.printStack()

    println("\n-- Clearing the Stack --")
    stack.clear()

    println("\n-- Checking if Stack is Empty --")
    println("Is stack empty? ${stack.isEmpty()}") // Should print true
}